core/src/OpenACC/Kokkos_OpenACC_ParallelScan_Range.hpp

```diff
@@ -0,0 +1,215 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_SCAN_RANGE_HPP
+#define KOKKOS_OPENACC_PARALLEL_SCAN_RANGE_HPP
+
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+#include <Kokkos_Parallel.hpp>
+
+namespace Kokkos::Experimental::Impl {
+template <class IndexType, class Functor, class ValueType>
+void OpenACCParallelScanRangePolicy(IndexType begin, IndexType end,
+                                    Functor afunctor, ValueType init_value,
+                                    int async_arg) {
+  auto const functor(afunctor);
+  const IndexType N                  = end - begin;
+  constexpr IndexType chunk_size     = 128;
+  constexpr IndexType log_chunk_size = 7;
+  const IndexType n_chunks           = (N + chunk_size - 1) / chunk_size;
+  const IndexType nteams             = n_chunks > 512 ? 512 : n_chunks;
+  Kokkos::View<ValueType*, Kokkos::Experimental::OpenACC> chunk_values(
+      "chunk_values", n_chunks);
+  Kokkos::View<ValueType*, Kokkos::Experimental::OpenACC> offset_values(
+      "offset_values", n_chunks);
+  ValueType element_values[2][chunk_size];
+
+#pragma acc enter data copyin(functor) copyin(chunk_values, offset_values) \
+    async(async_arg)
+
+#pragma acc parallel loop gang num_gangs(nteams)      \
+    vector_length(chunk_size) private(element_values) \
+        present(functor, chunk_values)
+  for (IndexType team_id = 0; team_id < n_chunks; ++team_id) {
+#pragma acc loop vector
+    for (IndexType thread_id = 0; thread_id < chunk_size; ++thread_id) {
+      const IndexType local_offset = team_id * chunk_size;
+      const IndexType idx          = local_offset + thread_id;
+      ValueType update             = init_value;
+      if ((idx > 0) && (idx < N)) functor(idx - 1, update, false);
+      element_values[0][thread_id] = update;
+    }
+    IndexType step_size    = 1;
+    IndexType current_step = 0;
+    IndexType next_step    = 1;
+    IndexType temp;
+    for (IndexType i = 0; i < log_chunk_size; ++i) {
+      if (i == 0) {
+        step_size = 1;
+      } else {
+        step_size *= 2;
+      }
+#pragma acc loop vector
+      for (IndexType thread_id = 0; thread_id < chunk_size; ++thread_id) {
+        if (thread_id < step_size) {
+          element_values[next_step][thread_id] =
+              element_values[current_step][thread_id];
+        } else {
+          element_values[next_step][thread_id] =
+              element_values[current_step][thread_id] +
+              element_values[current_step][thread_id - step_size];
+        }
+      }
+      temp         = current_step;
+      current_step = next_step;
+      next_step    = temp;
```

Comment:
This only makes `temp`, `current_step`, and `next_step`, alternate between `0,1,0` and `1,0,1`. Is that intended?

Reply:
Yes, the main purpose of these instructions is to toggle values of `current_step` and `next_step` between `0` and `1`. We can use other methods such as direct calculation (e.g., `current_step = 1 - current_step; next_step = 1 - next_step;`), but I guess copying between two local variables, which will be on registers, is the cheapest way.